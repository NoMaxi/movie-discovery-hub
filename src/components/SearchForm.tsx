import React, { useState, FormEvent, ChangeEvent } from "react";

interface SearchFormProps {
    onSearch: (query: string) => void;
    initialSearchQuery?: string;
}

const SearchForm = ({ initialSearchQuery = "", onSearch }: SearchFormProps) => {
    const [query, setQuery] = useState(initialSearchQuery);

    const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
        setQuery(event.target.value);
    };

    const handleSubmit = (event: FormEvent) => {
        event.preventDefault();
        onSearch(query);
    };

    const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
        if (event.key === "Enter") {
            handleSubmit(event);
        }
    };

    return (
        <form onSubmit={ handleSubmit } className="flex items-center justify-center">
            <input
                type="text"
                value={ query }
                onChange={ handleInputChange }
                onKeyDown={ handleKeyDown }
                placeholder="What do you want to watch?"
                className="
                    input-field
                    placeholder-[var(--color-gray-lighter)]
                    border-none
                    rounded-md
                    py-3
                    px-6
                    max-w-1/2
                    text-lg
                    flex-grow
                "
            />
            <button type="submit" className="btn ml-3">Search</button>
        </form>
    );
};

export default SearchForm;
