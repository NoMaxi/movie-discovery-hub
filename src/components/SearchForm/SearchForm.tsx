import React, { useState, FormEvent, ChangeEvent } from "react";

interface SearchFormProps {
    onSearch: (query: string) => void;
    initialQuery?: string;
}

const SearchForm = ({ initialQuery = "", onSearch }: SearchFormProps) => {
    const [query, setQuery] = useState(initialQuery);

    const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
        setQuery(event.target.value);
    };

    const handleSubmit = (event: FormEvent) => {
        event.preventDefault();
        const trimmedQuery = query.trim();
        if (trimmedQuery) {
            onSearch(trimmedQuery);
        }
    };

    const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
        if (event.key === "Enter") {
            handleSubmit(event);
        }
    };

    return (
        <form onSubmit={handleSubmit} className="flex items-center justify-center w-1/2">
            <input
                data-testid="search-input"
                type="text"
                value={query}
                onChange={handleInputChange}
                onKeyDown={handleKeyDown}
                placeholder="What do you want to watch?"
                className="
                    input-field
                    placeholder-[var(--color-gray-lighter)]
                    border-none
                    rounded-md
                    py-3
                    px-6
                    text-lg
                    flex-grow
                "
            />
            <button data-testid="search-button" type="submit" className="btn ml-3" disabled={!query.trim()}>
                Search
            </button>
        </form>
    );
};

export default SearchForm;
