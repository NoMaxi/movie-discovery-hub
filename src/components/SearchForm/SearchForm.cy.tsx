/// <reference types="@testing-library/cypress" />
import { SearchForm } from "./SearchForm";

describe("<SearchForm /> Component Tests", () => {
    const searchText = "movie title";
    const placeholderText = "What do you want to watch?";
    const searchButtonSelector = "[data-testid='search-button']";
    const searchInputSelector = "[data-testid='search-input']";

    it("Should render input with an initial value if initialQuery is provided in props", () => {
        const initialQuery = "test movie";
        const onSearchSpy = cy.spy();
        cy.mount(<SearchForm onSearch={onSearchSpy} initialQuery={initialQuery} />);

        cy.get(searchInputSelector).should("have.value", initialQuery);
    });

    it("Should render empty input if no initialQuery is provided in props", () => {
        const onSearchSpy = cy.spy();
        cy.mount(<SearchForm onSearch={onSearchSpy} />);

        cy.findByPlaceholderText(placeholderText).should("be.visible").and("have.value", "");
        cy.get(searchButtonSelector).should("be.visible").and("contain.text", "Search");
    });

    it("Should update input value when user types text in the input", () => {
        const onSearchSpy = cy.spy();
        cy.mount(<SearchForm onSearch={onSearchSpy} />);

        cy.get(searchInputSelector).type(searchText);

        cy.get(searchInputSelector).should("have.value", searchText);
    });

    it("Should call 'onSearch' prop with input value when search button is clicked", () => {
        const onSearchSpy = cy.spy().as("onSearchSpy");
        cy.mount(<SearchForm onSearch={onSearchSpy} />);

        cy.get(searchInputSelector).type(searchText);
        cy.get(searchButtonSelector).click();

        cy.get("@onSearchSpy").should("have.been.calledOnceWith", searchText);
    });

    it("Should call 'onSearch' prop with input value when Enter key is pressed in the input", () => {
        const onSearchSpy = cy.spy().as("onSearchSpy");
        cy.mount(<SearchForm onSearch={onSearchSpy} />);

        cy.get(searchInputSelector).type(`${searchText}{enter}`);

        cy.get("@onSearchSpy").should("have.been.calledOnceWith", searchText);
    });

    it("Should not call 'onSearch' prop when Enter key is pressed in the input if the input value is empty", () => {
        const onSearchSpy = cy.spy().as("onSearchSpy");
        cy.mount(<SearchForm onSearch={onSearchSpy} />);

        cy.get(searchInputSelector).type(`{enter}`);

        cy.get("@onSearchSpy").should("not.have.been.called");
    });

    it("Should not call 'onSearch' prop when a non-Enter key is pressed in the input", () => {
        const onSearchSpy = cy.spy().as("onSearchSpy");
        cy.mount(<SearchForm onSearch={onSearchSpy} />);

        cy.get(searchInputSelector).type(`${searchText}{downArrow}`);

        cy.get("@onSearchSpy").should("not.have.been.called");
    });
});
