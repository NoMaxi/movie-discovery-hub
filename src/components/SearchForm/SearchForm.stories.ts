import type { Meta, StoryObj } from "@storybook/react";
import { fn } from "@storybook/test";
import SearchForm from "./SearchForm";

const meta = {
    component: SearchForm,
    args: {
        onSearch: fn(),
    },
} satisfies Meta<typeof SearchForm>;

export default meta;

type Story = StoryObj<typeof meta>;

export const WithoutInitialQuery: Story = {
    args: {},
};

export const WithInitialQuery: Story = {
    args: {
        initialQuery: "Star Wars",
    },
};
