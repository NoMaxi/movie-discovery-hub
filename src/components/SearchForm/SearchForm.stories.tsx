import type { Meta, StoryObj } from "@storybook/react";
import { fn } from "@storybook/test";
import SearchForm from "./SearchForm";

const meta = {
    component: SearchForm,
    argTypes: {
        initialQuery: {
            control: "text",
            description: "Initial search query to be displayed in the input field",
        },
        onSearch: {
            control: false,
            description: "Callback function triggered when search is submitted",
        },
    },
    args: {
        onSearch: fn(),
    },
} satisfies Meta<typeof SearchForm>;

export default meta;

type Story = StoryObj<typeof SearchForm>;

export const WithoutInitialQuery: Story = {
    args: {
        initialQuery: "",
    },
};

export const WithInitialQuery: Story = {
    args: {
        initialQuery: "Star Wars",
    },
};

export const LongQuery: Story = {
    args: {
        initialQuery: "This is a very long search query that should fit the input field anyway",
    },
};
