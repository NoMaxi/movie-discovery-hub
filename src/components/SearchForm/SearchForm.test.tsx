import { render, fireEvent, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { SearchForm } from "./SearchForm";

describe("SearchForm", () => {
    const placeholderText = "What do you want to watch?";
    const searchText = "movie title";
    let onSearchMock: jest.Mock;

    beforeEach(() => {
        onSearchMock = jest.fn();
    });

    test("Should render input with an initial value if initialQuery is provided in props", () => {
        const { asFragment } = render(<SearchForm initialQuery={searchText} onSearch={onSearchMock} />);
        const input = screen.getByPlaceholderText(placeholderText);

        expect(input).toBeInTheDocument();
        expect(input).toHaveValue(searchText);
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should render empty input if no initialQuery is provided in props", () => {
        const { asFragment } = render(<SearchForm onSearch={onSearchMock} />);
        const input = screen.getByPlaceholderText(placeholderText);

        expect(input).toBeInTheDocument();
        expect(input).toHaveValue("");
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should update input value when user types text in the input", () => {
        render(<SearchForm onSearch={onSearchMock} />);
        const input = screen.getByPlaceholderText(placeholderText);

        fireEvent.change(input, { target: { value: searchText } });

        expect(input).toHaveValue(searchText);
    });

    test("Should call 'onSearch' prop with input value when search button is clicked", () => {
        render(<SearchForm onSearch={onSearchMock} />);
        const input = screen.getByPlaceholderText(placeholderText);
        const button = screen.getByText("Search");

        fireEvent.change(input, { target: { value: searchText } });
        fireEvent.click(button);

        expect(onSearchMock).toHaveBeenCalledWith(searchText);
    });

    test("Should call 'onSearch' prop with input value when Enter key is pressed in the input", () => {
        render(<SearchForm onSearch={onSearchMock} />);
        const input = screen.getByPlaceholderText(placeholderText);

        fireEvent.change(input, { target: { value: searchText } });
        fireEvent.keyDown(input, { key: "Enter" });

        expect(onSearchMock).toHaveBeenCalledWith(searchText);
    });

    test("Should not call 'onSearch' prop when Enter key is pressed in the input if the input value is empty", () => {
        render(<SearchForm onSearch={onSearchMock} />);
        const input = screen.getByPlaceholderText(placeholderText);

        fireEvent.keyDown(input, { key: "Enter" });

        expect(onSearchMock).not.toHaveBeenCalled();
    });

    test("Should not call 'onSearch' prop when a non-Enter key is pressed in the input", () => {
        render(<SearchForm onSearch={onSearchMock} />);
        const input = screen.getByPlaceholderText(placeholderText);

        fireEvent.change(input, { target: { value: searchText } });
        fireEvent.keyDown(input, { key: "Space" });

        expect(onSearchMock).not.toHaveBeenCalled();
    });
});
