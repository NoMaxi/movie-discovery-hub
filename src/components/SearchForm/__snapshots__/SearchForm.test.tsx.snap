// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm Should render empty input if no initialQuery is provided in props 1`] = `
<DocumentFragment>
  <form
    class="flex items-center justify-center w-1/2"
  >
    <input
      class="
                    input-field
                    placeholder-[var(--color-gray-lighter)]
                    border-none
                    rounded-md
                    py-3
                    px-6
                    text-lg
                    flex-grow
                "
      data-testid="search-input"
      placeholder="What do you want to watch?"
      type="text"
      value=""
    />
    <button
      class="btn ml-3"
      data-testid="search-button"
      disabled=""
      type="submit"
    >
      Search
    </button>
  </form>
</DocumentFragment>
`;

exports[`SearchForm Should render input with an initial value if initialQuery is provided in props 1`] = `
<DocumentFragment>
  <form
    class="flex items-center justify-center w-1/2"
  >
    <input
      class="
                    input-field
                    placeholder-[var(--color-gray-lighter)]
                    border-none
                    rounded-md
                    py-3
                    px-6
                    text-lg
                    flex-grow
                "
      data-testid="search-input"
      placeholder="What do you want to watch?"
      type="text"
      value="movie title"
    />
    <button
      class="btn ml-3"
      data-testid="search-button"
      type="submit"
    >
      Search
    </button>
  </form>
</DocumentFragment>
`;
